neuralop/layers/codano.py

```diff
@@ -0,0 +1,607 @@
+        q = q.view(batch_size, self.n_head, t, d * q.size(-2)
+                   * q.size(-1))
+        v = v.view(batch_size, self.n_head, t, d * v.size(-2)
+                   * v.size(-1))
+
+        dprod = (torch.matmul(q, k.transpose(-1, -2)) /
+                 (np.sqrt(k.shape[-1]) * self.temperature))
+        dprod = F.softmax(dprod, dim=-1)
+
+        attention = torch.matmul(dprod, v)
+
+        # Reshape from (b, a, t, d * h * w) to (b, a, t, d, h, w)
+        attention = attention.view(
+            attention.size(0),
+            attention.size(1),
+            attention.size(2),
+            self.head_codimension,
+            value_x,
+            value_y)
+
+        # Permute the dimensions to rearrange into (b * t, a * d, h, w)
+        attention = attention.permute(
+            0, 2, 1, 3, 4, 5).contiguous()  # (b, t, a, d, h, w)
+
+        # Reshape to flatten the first and second dimensions (b * t) and the
+        # third and fourth dimensions (a * d)
+        attention = attention.view(attention.size(0) * attention.size(1),
+                                   attention.size(2) * attention.size(3),
+                                   value_x, value_y)  # (b * t, a * d, h, w)
+
+        return attention
+
+    def forward(self, x):
+        if self.permutation_eq:
+            return self._forward_equivariant(x)
+        else:
+            return self._forward_non_equivariant(x)
+
+    def _forward_equivariant(self, x):
+        batch_size = x.shape[0]
+        output_shape = x.shape[-2:]
+
+        assert x.shape[1] % self.token_codimension == 0
+
+        # reshape 'b (t d) h w -> (b t) d h w'
+        t = x.size(1) // self.token_codimension
+        xa = x.view(
+            x.size(0),
+            t,
+            self.token_codimension,
+            x.size(2),
+            x.size(3))  # (b, t, d, h, w)
+
+        xa = xa.permute(0, 1, 2, 3, 4).contiguous()
+        xa = xa.view(
+            x.size(0) * t,
+            self.token_codimension,
+            x.size(2),
+            x.size(3))  # (b * t, d, h, w)
+        xa_norm = self.norm1(xa)
+        attention = self.compute_attention(xa_norm, batch_size)
+        if self.proj is not None:
+            attention = self.proj(attention)
+
+        attention = self.attention_normalizer(attention + xa)
+        attention_normalized = self.norm2(attention)
+        output = self.mixer(attention_normalized, output_shape=output_shape)
+
+        output = self.mixer_out_normalizer(output) + attention
+
+        # reshaped '(b t) d h w -> b (t d) h w'
+        t = output.size(0) // batch_size
+        output = output.view(
+            batch_size,
+            t,
+            output.size(1),
+            output.size(2),
+            output.size(3))
+        output = output.view(
+            batch_size,
+            t * output.size(2),
+            output.size(3),
+            output.size(4))
+
+        return output
+
+    def _forward_non_equivariant(self, x):
```

Comment:
Would be good to add docstrings explaining the difference between different forward call types

Reply:
Added.